### PR TITLE
version_info.log  added in the log directory

### PR DESCRIPTION
**Pull request by mpompolas**
version_info.log holds information about the version of ivadomed toolbox, the Git-Annex of the dataset used, system info and user inputs.

Example:

```
IVADOMED TOOLBOX
----------------
(git-kn/versions_log-6b371894cd202a9d134c07e20be281f0bc1a9dc8*)


DATASET VERSION
---------------
A single BIDS dataset was used for training.
/home/nas/ivado-project/Datasets/data_example_spinegeneric/ - Dataset Annex version: e0cebf8949281761a325c110ab84384e37e697f7*


SYSTEM INFO
-------------
OS: linux (Linux-5.8.14-050814-generic-x86_64-with-glibc2.10)
CPU cores: Available: 16


CONFIG INPUTS
-------------
command: test
gpu: 0
log_directory: /home/nas/PycharmProjects/ivadomed-personal-scripts/log_to_delete
model_name: seg_sc_t1-t2-t2s-mt
debugging: False
object_detection_params: {'object_detection_path': None, 'safety_factor': [1.0, 1.0, 1.0]}
loader_parameters: {'bids_path': '/home/nas/ivado-project/Datasets/data_example_spinegeneric/', 'subject_selection': {'n': [], 'metadata': [], 'value': []}, 'target_suffix': ['_seg-manual'], 'extensions': [], 'roi_params': {'suffix': None, 'slice_filter_roi': None}, 'contrast_params': {'training_validation': ['T1w', 'T2w', 'T2star'], 'testing': ['T1w', 'T2w', 'T2star'], 'balance': {}}, 'slice_filter_params': {'filter_empty_mask': False, 'filter_empty_input': True}, 'slice_axis': 'axial', 'multichannel': False, 'soft_gt': False}
split_dataset: {'fname_split': None, 'random_seed': 6, 'center_test': ['mpicbs'], 'method': 'per_center', 'balance': None, 'train_fraction': 0.6, 'test_fraction': 0.2}
training_parameters: {'batch_size': 18, 'loss': {'name': 'DiceLoss'}, 'training_time': {'num_epochs': 2, 'early_stopping_patience': 50, 'early_stopping_epsilon': 0.001}, 'scheduler': {'initial_lr': 0.001, 'lr_scheduler': {'name': 'CosineAnnealingLR', 'base_lr': 1e-05, 'max_lr': 0.01}}, 'balance_samples': {'applied': False, 'type': 'gt'}, 'mixup_alpha': None, 'transfer_learning': {'retrain_model': None, 'retrain_fraction': 1.0, 'reset': True}}
default_model: {'name': 'Unet', 'dropout_rate': 0.3, 'bn_momentum': 0.9, 'depth': 3, 'is_2d': True}
uncertainty: {'epistemic': False, 'aleatoric': False, 'n_it': 0}
postprocessing: {'remove_noise': {'thr': -1}, 'keep_largest': {}, 'binarize_prediction': {'thr': 0.5}, 'uncertainty': {'thr': -1, 'suffix': '_unc-vox.nii.gz'}, 'fill_holes': {}, 'remove_small': {'unit': 'vox', 'thr': 3}}
evaluation_parameters: {'targetSize': {'unit': 'vox', 'thr': [20, 100]}, 'overlap': {'unit': 'vox', 'thr': 3}}
transformation: {'Resample': {'wspace': 0.75, 'hspace': 0.75, 'dspace': 1}, 'CenterCrop': {'size': [128, 128]}, 'RandomAffine': {'degrees': 5, 'scale': [0.1, 0.1], 'translate': [0.03, 0.03], 'applied_to': ['im', 'gt'], 'dataset_type': ['training']}, 'ElasticTransform': {'alpha_range': [28.0, 30.0], 'sigma_range': [3.5, 4.5], 'p': 0.1, 'applied_to': ['im', 'gt'], 'dataset_type': ['training']}, 'NumpyToTensor': {}, 'NormalizeInstance': {'applied_to': ['im']}}
FiLMedUnet: {'applied': False, 'metadata': 'contrasts', 'film_layers': [0, 1, 0, 0, 0, 0, 0, 0]}
Modified3DUNet: {'applied': False, 'length_3D': [128, 128, 16], 'stride_3D': [128, 128, 16], 'attention': False, 'n_filters': 8}
```

### CODE REVIEW

File: ivadomed/main.py [hunk 1]
```diff
@@ -378,6 +384,75 @@ def run_command(context, n_gif=0, thr_increment=None, resume_training=False):
...
+        f.write('The following BIDS datasets were used for training.\n')
+
+        for iDataset in len(dataset_paths):
+            if len(datasets_version[0]) != 0:
+                f.write(str(iDataset+1) + '. ' + dataset_paths[iDataset] + ' - Dataset Annex version: ' + datasets_version[0] + '\n')
+            else:
+                f.write('\n')
+
+    # SYSTEM INFO
+    f.write('\n\nSYSTEM INFO\n-------------\n')
```
Comment (ahill187) [inline ivadomed/main.py:429]: I know we don't officially support Windows at the moment due to some bugs, should we add a check for `os_running = 'windows'` for when we add Windows support in the future? Or cross that bridge when we get there?

File: ivadomed/main.py [hunk 2]
```diff
@@ -378,6 +384,75 @@ def run_command(context, n_gif=0, thr_increment=None, resume_training=False):
...
+        for Dataset in dataset_paths:
+            datasets_version.append(imed_utils.__get_commit(path_to_git_folder=Dataset))
+
+    log_file = os.path.join(context['log_directory'], 'version_info.log')
+
+    try:
+        f = open(log_file, "w")
+    except OSError as err:
+        print("OS error: {0}".format(err))
+        raise Exception("Have you selected a log folder you have writing permissions?")
```
Comment (ahill187) [inline ivadomed/main.py:406]: Maybe change this to ```raise Exception("Have you selected a log folder, and do you have write permissions for that folder?")```?

**Review by ahill187 — COMMENTED** [review]
I like this idea, super helpful!! The log file looks great.

File: ivadomed/main.py (hunk 1 above)
Comment (mpompolas) [reply ivadomed/main.py:429]: yes we should!
Thanks for the suggestion

Comment (ahill187) [reply ivadomed/main.py:429]: No problem!!

**Review by ahill187 — APPROVED** [review]

**Review by charleygros — APPROVED** [review]
Love it!

**Pull request merged**